oak/server/rust/oak_runtime/src/runtime/mod.rs

```diff
@@ -169,25 +187,124 @@ impl Runtime {
         // for any additional work to be finished here. This may take an arbitrary amount of time,
         // depending on the work that the node thread has to perform, but at least we know that the
         // it will not be able to enter again in a blocking state.
-        for (_, mut node) in nodes.drain() {
-            node.instance.stop();
+        for mut instance in instances.drain(..) {
+            instance.stop();
+        }
+    }
+
+    /// Allow the corresponding [`Node`] to use the [`Handle`]s passed via the iterator.
+    /// This is achieved by adding the [`Handle`]s to the [`Node`]s [`HashMap`] of [`Handle`]s.
+    fn track_handles_in_node<'a, I: Iterator<Item = &'a Handle>>(
+        &self,
+        node_id: NodeId,
+        handles: I,
+    ) {
+        if node_id == RUNTIME_NODE_ID {
+            return;
+        }
+
+        let nodes = self.nodes.read().unwrap();
+        if let Some(node) = nodes.get(&node_id) {
+            let mut tracked_handles = node.handles.lock().unwrap();
+            for handle in handles {
+                tracked_handles.insert(*handle);
+            }
+        }
+    }
+
+    /// Replace [`Handle`] with [`None`] if the [`NodeId`] does not have access to the [`Handle`].
+    fn filter_handle(&self, node_id: NodeId, handle: Handle) -> Option<Handle> {
```

Comment:
The caller still has access to the handle, so returning it again wrapped in an `Option` seems a bit odd. Perhaps this should just return a `bool` and be renamed accordingly (e.g. `can_access_handle`)?